download: some fixes

mcprat opened the pull request:
this is a small part of #9975

mcprat committed:
download: improve svn download method

The previous option --trust-server-cert is marked deprecated
and has a new option in place,
for different ways a cert can be trusted by override.

Try secure download before insecure download.

Delete the target file before replacing it,
which seems to help a bug related to NFS (network filesystem).

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
scripts/download.pl: ignore target file if empty

Let the target file be hashed or replaced
based on its size instead of just existing,
for compatibility with flock.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
scripts/download.pl: improve file cleanup

Organize the removal of the target file into a function, trash()

Make more error cases end with deletion of the target file
instead of just removing the hash and tmp file.

When hash does not match, the target file is supposed to be deleted,
so place that line before the "die" message.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
download: fix integrity check for host build

Download/Default must be called before the integrity check,
otherwise nothing happens if the file exists but hash differs
for tools / toolchain / other host-compile calls.

Rename the recipe to match the naming style
in the rest of the makefile.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/cmake: fix download url with make variables

Use a make variable pattern for the url
so that only one version number needs to be changed
when version is bumped.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

neheb commented on scripts/download.pl:
```diff
@@ -284,7 +284,7 @@ sub trash
 		exit 0 if $sum eq $file_hash;
 
-		die "Hash of the local file $filename does not match (file: $sum, requested: $file_hash) - deleting download.\n";
+		warn "Hash of the local file $filename does not match (file: $sum, requested: $file_hash) - deleting download.\n";
```
NACK

mpratt14 replied:
tell me your thoughts on it

AFAIK, it just tries the other mirrors too and then dies the same way

if I were to touch the filename to be downloaded and then run the download it will only exit without attempting a download

mpratt14 replied:
by the way, the script dies saying "deleting download" before the line where it deletes the file, so it doesn't even do what it says it does

Ansuel replied:
> by the way, the script dies saying "deleting download" before the line where it deletes the file, so it doesn't even do what it says it does

ok this is actually a bug and we should take more care of the dl dir... (but keeping the affected file would help us bisect the problem)

Anyway I assume NACK because having different hash between a local file and a remote file is a hard error and something really bad is happening... aka REAL SECURITY CONCERN.

If someone notice this kind of thing the first question to do would be "Is my system compromised? Is the remote system compromised? Did the makefile just have a different hash?"

Hash values are not string values just to check the integrity of the file but also to check the integrity of the entire system both client and remote.

So yes I agree... I would keep the `die` since having mismatched hash is REALLY BAD. (and just ignoring it as we have other mirror is not a good idea as (thinking absurdly) we would keep a compromised mirror and never notice that)

mpratt14 replied:
I agree it would be considered a security concern, and should error out the script, the only difference is when

it would print hash mismatch, delete the file, and then print out of mirrors and die there

if there is a compromised file on a mirror, or a compromised mirror, and its not the first mirror in the mirrors list, the script would never tell us anyway, because the first attempt would be OK.

if the client is compromised, every download attempt fails, even with "warn" here, because another part of the script handles hashing the download apart from the local file if it is already there. (funny enough, I broke mkhash during testing and had that exact situation)

A larger question is whether we want to keep the local file or not if hash does not match, and if so, only the message in the die should change. However, if we REALLY consider it a security concern, I would say it should get deleted and a new download attempt happens immediately, which is what "warn" would do (but again, I understand that does assume the local file got ruined randomly and the remote file and client are not compromised, which is not likely to occur with normal usage)

the use case I actually care about is for the script to not freak out and error like there is a security concern if the file is empty, and I realize now changing from "die" to "warn" is not the only way to handle that case

I leave that decision of which change is better up to you

Ansuel replied:
Thing is what are the possibility of having an empty file?

If we really want we can add extra check on mismatched hash (like check for the remote file size and check if ours match... and then hard exit...)

To me your concern is more on checking if a file is actually corrupt and that makes sense but how common is that?

I would honestly just fix the bug with not deleting the file and be done with it :D lets not waste too much time on small thing, you already have monster pr to split ahhahah

neheb commented on tools/cmake/Makefile:
```diff
@@ -8,12 +8,13 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=cmake
 PKG_VERSION:=3.22.3
+PKG_VERSION_MAJOR:=$(word 1,$(subst ., ,$(PKG_VERSION))).$(word 2,$(subst ., ,$(PKG_VERSION)))
```
ACK

mpratt14 commented:
ok I'm keeping things very simple now 👍🏼

mpratt14 commented:
if you dont like 79fdc3c605d2028332bdb0c303bb9e054d81efd2 we just get rid of it, it's insignificant compared to the rest

Ansuel commented:
@mpratt14 nha it's ok just the commit description require a little bit of care like Capitalize words, dot at the end of the sentence. I will check the commit asap...

Ansuel commented:
ok @mpratt14 just checked the commits, yes i confirm you should fix the commit description and for download.pl changes use
'scripts/download.pl' instead of download.pl :D thanks for the pr

Ansuel commented:
Running this under https://github.com/Ansuel/openwrt/actions/runs/2683547470 and if everything goes well (excluding my bad ci  that is still WIP) I will merge this. :D

mpratt14 commented:
@Ansuel do you think this is ready by the way?

mpratt14 commented:
@Ansuel ping

I believe the only error in your Github action is unrelated to this

mpratt14 commented:
@Ansuel you think this part is ready to merge? then I split another few commits from the large one

jow- commented on tools/cmake/Makefile:
```diff
@@ -8,12 +8,13 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=cmake
 PKG_VERSION:=3.22.3
+PKG_VERSION_MAJOR:=$(word 1,$(subst ., ,$(PKG_VERSION))).$(word 2,$(subst ., ,$(PKG_VERSION)))
```
`$(wordlist 1,2,$(subst ., ,$(PKG_VERSION)))` could be a more concise variant that does not require substituting twice.

mpratt14 replied:
hmmmm, but the delimiter must be returned to a dot instead of a space...

jow- replied:
Ah right, of course. Nevermind then.

mpratt14 commented:
@hauke this should be ready for merge, or ask Ansuel if not sure

Ansuel commented:
Sorry for the delay, I was on holiday... Will check this if Hauke doesn't
do it early

Il Lun 1 Ago 2022, 14:12 Michael Pratt ***@***.***> ha
scritto:

> @hauke <https://github.com/hauke> this should be ready for merge, or ask
> Ansuel if not sure
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/10250#issuecomment-1201121594>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQQGVJDTXCWMN2M73QTVW65LNANCNFSM53JLJQNQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mcprat committed:
rules.mk: add option to download files into a subdirectory

Some packages download a single raw file or set of raw files
instead of an archive.

The filename of many of these files have names and version numbers
that do not directly correspond to the name and version of the package
as it is defined in its Makefile.

To better organize downloaded files, this adds the option of
creating a subdirectory in the download directory for these files
that can be named with package specific variables.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
@Ansuel I have one last little commit that fits the scope of this PR that I just put in #9975 (HEAD commit) to see the CI results

if it's good I'm going to move it here

mpratt14 commented:
added the commit that I mentioned, it required rebase, no change to the other commits

mpratt14 commented:
something weird happening to the CI

```
Cloning into './feeds/luci'...
error: RPC failed; HTTP 504 curl 22 The requested URL returned error: 504 Gateway Time-out
fatal: the remote end hung up unexpectedly
failed.
```

mpratt14 commented:
rebased for the conflict

for the last commit, only using PKG_NAME and PKG_VERSION in folder name

@Ansuel ready for final review

mpratt14 commented:
@Ansuel @hauke @aparcar can we do final review and merge?

Ansuel commented on package/firmware/ath10k-ct-firmware/Makefile:
```diff
@@ -1,9 +1,11 @@
 PKG_VERSION:=2020-11-08
 PKG_RELEASE:=1
 
+DL_MKDIR:=1
```
Should we document this in the wiki?

mpratt14 replied:
I suppose, but I hope the commit message is self-explanatory enough

Ansuel replied:
I personally first check the makefile wiki page to search if a special magic value can do what I need so yes adding this to the wiki is mandatory :D

Ansuel commented:
@mpratt14 I know you will hate me and sorry for the request...

Can you split the last commit in:
- commit that introduce the new variable and put an example in the commit
- one commit for each package that you change?

I ask you to do this to have better track of the changes and because in some package you do more change than simply adding the new variable.

Aside from that everything else looks good to me

mcprat committed:
acx-mac80211: set DL_DIR to package specific subdirectory

This package downloads raw files
which have names that are not corresponding to
the name and version of the package
as it is defined in the Makefile.

Use the option DL_SUBDIR to set the DL_DIR
to be a subdirectory named with
PKG_NAME and PKG_VERSION
to better organize the downloads.

Define PKG_VERSION here
using PKG_SOURCE_DATE.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
ath10k-ct-firmware: set DL_DIR to package specific subdirectory

This package downloads raw files
which have names that are not corresponding to
the name and version of the package
as it is defined in the Makefile.

Use the option DL_SUBDIR to set the DL_DIR
to be a subdirectory named with
PKG_NAME and PKG_VERSION
to better organize the downloads.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
b43legacy-firmware: set DL_DIR to package specific subdirectory

This package downloads raw files
which have names that are not corresponding to
the name and version of the package
as it is defined in the Makefile.

Use the option DL_SUBDIR to set the DL_DIR
to be a subdirectory named with
PKG_NAME and PKG_VERSION
to better organize the downloads.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
prism54-firmware: set DL_DIR to package specific subdirectory

This package downloads raw files
which have names that are not corresponding to
the name and version of the package
as it is defined in the Makefile.

Use the option DL_SUBDIR to set the DL_DIR
to be a subdirectory named with
PKG_NAME and PKG_RELEASE
to better organize the downloads.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
fortunately, I have learned enough of git at this point so that request is not difficult 👍🏼 

and yeah I get the reason, each package needs separate commit with it's name in the title

mpratt14 commented:
wait a minute.... does ipq-wifi not download anything?

I can't remember what file I saw that made me add it

mpratt14 commented:
@Ansuel should it be called `DL_SUBDIR` or do we like `DL_MKDIR`

mpratt14 commented:
@Ansuel I saved the old way, but I think this way is better, giving more control to the individual package Makefile

ready for review again

mpratt14 commented:
I'm going to split this PR

the commit
`download: fix integrity check for host build`
is not ready

the problem is more complex than I thought, and it would be better handled if changes in another PR happens first

Ansuel commented:
@mpratt14 can you rebase?

mpratt14 commented:
@Ansuel right now this PR is loosely dependent on #10659 and I have to rewrite the `download: fix integrity check for host build`  commit, so I have to either bring commits from there to here or we handle that PR first (with or without the "tools" tarball build, or maybe splitting the other PR in two)

mpratt14 commented:
...and the reason is that I realized while working on #10659 that the current approach in commit `download: fix integrity check for host build` is not correct even though it does work
